src/librawspeed/common/DngOpcodes.cpp

```diff
@@ -1,698 +1,424 @@
-        for (uint64 p = 0; p < mPlanes; p++)
-        {
-          src[x*cpp+p] = delta * src[x*cpp+p];
-        }
-      }
+public:
+  OffsetPerRowOrCol(ByteStream& bs) : DeltaRowOrColBase(bs, 65535.0f) {}
+
+  void apply(RawImage& ri) override
+  {
+    if (ri->getDataType() == TYPE_USHORT16) {
+      applyOP<ushort16>(ri, [this](uint32 x, uint32 y, ushort16 v) {
+        return clampBits(deltaI[S::select(x, y)] + v, 16);
+      });
+    } else {
+      applyOP<float>(ri, [this](uint32 x, uint32 y, float v) {
+        return deltaF[S::select(x, y)] + v;
+      });
     }
   }
-}
+};
 
-/***************** OpcodeScalePerCol   ****************/
+using OffsetPerRow = OffsetPerRowOrCol<SelectY>;
+using OffsetPerCol = OffsetPerRowOrCol<SelectX>;
 
-OpcodeScalePerCol::OpcodeScalePerCol(const uchar8* parameters, uint32 param_max_bytes, uint32 *bytes_used )
+template <typename S>
+class ScalePerRowOrCol : public DeltaRowOrColBase
 {
-  if (param_max_bytes < 36)
-    ThrowRDE("OpcodeScalePerCol: Not enough data to read parameters, only %u bytes left.", param_max_bytes);
-  mAoi.setAbsolute(getLong(&parameters[4]), getLong(&parameters[0]), getLong(&parameters[12]), getLong(&parameters[8]));
-  mFirstPlane = getLong(&parameters[16]);
-  mPlanes = getLong(&parameters[20]);
-  mRowPitch = getLong(&parameters[24]);
-  mColPitch = getLong(&parameters[28]);
-  if (mPlanes == 0)
-    ThrowRDE("OpcodeScalePerCol: Zero planes");
-  if (mRowPitch == 0 || mColPitch == 0)
-    ThrowRDE("OpcodeScalePerCol: Invalid Pitch");
-
-  mCount = getLong(&parameters[32]);
-  *bytes_used = 36;
-  if (param_max_bytes < 36 + (mCount*4))
-    ThrowRDE("OpcodeScalePerCol: Not enough data to read parameters, only %u bytes left.", param_max_bytes);
-  if ((uint64)mAoi.getWidth() != mCount)
-    ThrowRDE("OpcodeScalePerCol: Element count (%llu) does not match width of area (%d).", mCount, mAoi.getWidth());
-
-  for (uint64 i = 0; i < mCount; i++)
-    mDelta[i] = getFloat(&parameters[36+4*i]);
-  *bytes_used += 4*mCount;
-  mFlags = MultiThreaded;
-  mDeltaX = nullptr;
-}
+public:
+  ScalePerRowOrCol(ByteStream& bs) : DeltaRowOrColBase(bs, 1024.0f) {}
 
-OpcodeScalePerCol::~OpcodeScalePerCol() {
-  if (mDeltaX)
-    delete[] mDeltaX;
-  mDeltaX = nullptr;
-}
+  void apply(RawImage& ri) override
+  {
+    if (ri->getDataType() == TYPE_USHORT16) {
+      applyOP<ushort16>(ri, [this](uint32 x, uint32 y, ushort16 v) {
+        return clampBits((deltaI[S::select(x, y)] * v + 512) >> 10, 16);
+      });
+    } else {
+      applyOP<float>(ri, [this](uint32 x, uint32 y, float v) {
+        return deltaF[S::select(x, y)] * v;
+      });
+    }
+  }
+};
 
+using ScalePerRow = ScalePerRowOrCol<SelectY>;
+using ScalePerCol = ScalePerRowOrCol<SelectX>;
 
-RawImage& OpcodeScalePerCol::createOutput( RawImage &in )
+// ****************************************************************************
+
+DngOpcodes::DngOpcodes(TiffEntry *entry)
```

Comment:
This OSX + stack usage situation is slightly annoying. You seem to have been struggling with it for a while already. This particular constructor hardly changed at all. The `sizeof(ByteStream)` is neglectable. Since there are only very select places where the mac-code differs at all: why not completely disable the error on mac and be happy as long as the linux builds don't blow a reasonable stack size limit?